### ISSUE TITLE
Add markdown or rich text support

### ISSUE THREAD

**Issue by jonpalmisc** [jonpalmisc/bn-notepad#2]
It would be nice to have markdown or rich text support in the text editor, which is currently plaintext. Markdown could be complicated, rich text might be easier, but I forget what Qt offers in regard to rich text support in `QTextEdit` widgets.

**Comment by jonpalmisc** [comment]
## Markdown support roadmap

So initial support for headings was just added in v0.2.0. The feature roadmap is now as shown below.

### Up next
- [ ] Bold text
- [ ] Italic text
- [ ] Mono text (inline code)
- [ ] Hyperlinks
- [ ] Internal hyperlinks (links to addresses, functions, etc.)

### In the future
- [ ] Numbered & bulleted lists (potentially w/ automatic indent)
- [ ] Code blocks

### Not planned
- Images
- Todo lists
- Block/inline quotes

**Comment by galenbwill** [comment]
I would lobby for todo lists -- this was a huge part of my usage for markdown notes on a bndb.

I would also ask for ~~strikeout~~ text and just block quotes.

Not to be difficult, these are all just things i found useful in my workflow.


But definitely +1 for the internal hyperlinks. Regular hyperlinks are tempting, esp if there is a way to link/load other markdown files, but that's getting silly.


All that being said, it looks great so far.

**Comment by jonpalmisc** [comment]
Thank you for the positive feedback, I will definitely keep these requests in mind moving forward.

Right now, I really need to evaluate how the editor is going to be implemented long term. Currently, it is implemented as a `QPlainTextEditor` combined with a `QSyntaxHighlighter` used to apply formatting to headers. As the scope of the project increases, a few questions come to mind:

1. **Should the Markdown "operators" be visible in the editor?** Right now, the # symbol used to denote headings is visible, and if bold text were to be implemented, the asterisks would be visible. Hiding these would be added effort and could get messy, so further consideration needs to be applied to determine if hiding them is worth it.
2. **Would a traditional rich text editor be better?** I was thinking that something like [MRichTextEditor](https://github.com/Anchakor/MRichTextEditor) might be a better long-term choice. This would reduce the speed at which text can be formatted (as it would require button clicks or keyboard shortcuts), but would require less custom functionality to make formatting, etc. work. Rather than trying to implement everything ourselves, we could make [Qt's rich text engine](https://doc.qt.io/qt-5/richtext-html-subset.html) do the bulk of the work for us. (This would solve pretty much everything short of TODO lists, which I'm sure could also be tacked on.)

I'm curious to hear what your thoughts are from a user perspective. @galenbwill

**Comment by galenbwill** [comment]
I looked at your editor and highlighting implementation, and can see a relatively straightforward path to support all the core markdown features without getting too heavyweight.

I prefer to see all the markdown syntax, and prefer to use markdown versus a rich text editor -- markdown affords structure and basic emphasis without having to interact with any additional GUI elements or keyboard shortcuts.

I've used [bear editor](https://bear.app) off and on for years on iOS and macOS, and I go back and forth on whether I like the way it renders the markdown but keeps the markdown syntax.

My primary markdown workflow is using vscode with the Enhanced Markdown Preview plugin, but I find that rather unsatisfying as well. But looking at that disjointed presentation in vscode I realize that what I like about the direction you're headed is the way the headings scale: something like bear.app's syntax-preserving inline rendering, but without getting too "cute".

One question is whether to stick with a monospace font, or use a proportional font except in code blocks and spans. I'd prefer monospace -- I think that's one of the things I _don't_ like about bear editor -- but then code spans would probably need to be colored differently to make them stand out.

**tl;dr** IMHO:

1. Stick with the plain(ish) text editor
2. Keep the markdown "operators" visible
3. Continue adding formatting to "render" the markdown in-place according to the operators
4. Keep using a monospace font overall, but color code spans to set them off from other text

**Comment by jonpalmisc** [comment]
Thank you for the valuable feedback. I'm going to keep weighing different options and opinions, but here are my thoughts for right now:

1. I think I'll likely stick to a plaintext editor.
2. Operators likely stay visible.
4. I see the argument for monospace/proportional font - it will be exposed as a setting in a future update.

Regarding point 3 - one other idea I had would be to split the notepad into edit & view modes. If I were to do this, I would forgo almost all formatting in the editor (except for maybe heading font sizes), and concentrate the formatting efforts on the "view" mode. Going to need to think about this bit for a while longer though.

**Comment by galenbwill** [comment]
IMHO, the split edit/view mode is less useful in the context of a binja notepad. Markdown is less interesting for its rendered presentation than for its structure here. It would be just as reasonable to use RST or org-mode, but markdown is my go-to for taking notes because it's familiar and it's a very simple way to get some of the structure of org-mode, but still have some "formatting" features for emphasis or for code blocks when necessary.

So I would continue to lobby for the current direction.

**Comment by mechanicalnull** [comment]
Having the ability to click or double-click a function name or address in order to navigate there in Binary Ninja is my top request. Unformatted blocks are nice to get monospacing would be the only thing I'd put on the list, but it's secondary.

**Comment by jonpalmisc** [comment]
Thank you for the input! Links are definitely top priority right now, and are being tracked under #8. I hope to get them working soon.

**Comment by jonpalmisc** [comment]
After lots of consideration, I have decided to:
- split notepad into separate viewer/editor tabs;
- use a monospace font for the editor tab; and
- remove all highlighting from the editor tab.

The viewer tab will leverage the recently added Markdown support in Qt that I didn't know about. As a result, the entire GitHub-flavored Markdown spec should be supported. Some of these changes are live in the `tabs` branch, and I hope to have an update out in the coming days. Closing this issue.
